keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,618 @@
+// side of the tunnel. The format is inspired by golang's net interface
+// https://golang.org/pkg/net/#Dial 
+message DialRequest {
+    // network representing a named network. "Tcp" is the only supported
+    // value
+    string network = 1;
+
+    // For TCP network, the address has the form "ip:port", where host must
+    // be IP address.
+    string address = 2;
+
+    // random is the randomly generated bytes that represents a given dial request.
+    // the number is kept unchanged across the proxies, and copied to DialResponse
+    // from the other side of the tunnel. The number cannot be reused across 
+    // different DialRequests.
+    bytes random = 3;
+}
+
+// DialResponse is the response to a DialRequest.
+message DialResponse {
+    // Error when the dial request cannot be fulfilled.
+    Error error = 1;
+
+    // connectID represents a unique connect ID for the connection/stream.
+    int64 connectID = 2;
+
+    // Copied from DialRequest. Explained also in DialRequest.
+    bytes random = 3;
+}
+
+// CloseRequest requests to close a connection
+message CloseRequest {
+    // connection id to close.
+    int64 connectID = 1;
+}
+
+// CloseResponse is the response to a close connection request.
+message CloseResponse {
+    // connection id to close.
+    int64 connectionID = 1;
+
+    // Error when close request cannot be fulfilled.
+    Error error = 2;
+}
+
+message Data {
+    // connectID that data or error payload is belonged to
+    int32 connectID = 1;
+
+    // data payload
+    bytes data = 2;
+
+    // Error that happens to the connection.
+    Error error = 3;
+}
+```
+
+### Connectivity Proxy
+
+The connectivity proxy(s) can run in the same container as the KAS, 
+should run on the same machine and must run on the same flat network as the KAS. 
+It would listen on one port for gRPC connections from the KAS. 
+This port would be for forwarding traffic to the appropriate cluster. 
+It should also expose an administrative port speaking https. 
+This port is primarily for serving the liveness probe. 
+Without the liveness probe we can easily end up with a partially broken cluster 
+where the KAS cannot perform any of the requests requiring connections from the KAS 
+to the cluster. In addition this port could be used for serving things 
+such as pprof debug commands and monitoring data for the proxy. 
+
+### Direct Connection
+
+This connection type option is to allow for the use of the default dialer 
+if we want to bypass the connectivity proxy. 
+Among other things it allows for the connectivity service to be used 
+without actually using the connectivity proxy. 
+Once we roll out the feature this is a quick way to run the system 
+in a “legacy” mode or as a fallback. It also allows for very simple clusters 
+which do not need the concept of network segregation to be run 
+without the overhead (either in extra hop latency or configuration complexity) 
+of using the connectivity proxy.
```

Comment:
Would this be the implicit configuration if proxies were not configured?  (Less secure, but I'm guessing so for compatibility)

Reply:
I think so yes.